feat(gRPC): add internal gRPC API including event subscription and filtering (experimental)

bingyanglin opened the pull request:
# Description of change

This PR contains a batch of experimental gRPC API implementations for internal usage.

## Links to any relevant issues

List of included PRs:
- #7511 
- #8019
- #7746
- #8371
- #8412 
- #8443

## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that new and existing unit tests pass locally with my changes

### Release Notes

- [ ] Protocol:
- [x] Nodes (Validators and Full nodes): add experimental "internal" gRPC API, including subscription to events with filtering
- [ ] Indexer:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:
- [x] Internal gRPC API: add gRPC streaming for checkpoint data and checkpoint summary, and use a trait to access the storage layer

bingyanglin committed:
feat(iota-core): add internal gRPC API for checkpoint fetching (experimental) (#7511)

Currently the indexer syncs the checkpoints via a REST-API via polling,
or via filesystem.

This PR adds an internal experimental gRPC API, that supports
subscriptions to get new checkpoints. This should reduce the latency
between checkpoint creation and indexing.

The profiling results by using a local-network (4 validator, 4 nodes, 2
indexers, 2 Postgres DBs):

![image](https://github.com/user-attachments/assets/73a935f1-96e0-4340-ac7a-226dbdb4e341)

More details and steps to reproduce the profiling results were
documented and discussed at #6995

- [x] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [x] I have added tests that prove my fix is effective or that my
feature works
- [ ] I have checked that new and existing unit tests pass locally with
my changes

- [ ] Protocol:
- [x] Nodes (Validators and Full nodes): add experimental "internal"
gRPC API
- [ ] Indexer:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:
- [x] internal gRPC API: add gRPC streaming for checkpoint data and
checkpoint summary

---------

Co-authored-by: muXxer <git@muxxer.de>

bingyanglin committed:
feat(node): use trait to access node Database in gRPC (#8019)

Use a trait to access the storage layer for gRPC.

Part of #7723

- [x] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [x] Patch-specific tests (correctness, functionality coverage)
- [x] I have added tests that prove my fix is effective or that my
feature works
- [x] I have checked that new and existing unit tests pass locally with
my changes

- [ ] Protocol:
- [ ] Nodes (Validators and Full nodes):
- [ ] Indexer:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:
- [x] gRPC API: use a trait to access the storage layer

---------

Co-authored-by: muXxer <git@muxxer.de>
Co-authored-by: Vlad Semenov <vlad.semenov@iota.org>

bingyanglin committed:
feat(gRPC): Add subscription to events including filtering (#7746)

# Description of change

Add event filters for client subscription

## Links to any relevant issues

fixes #7726

## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [x] I have added tests that prove my fix is effective or that my
feature works
- [x] I have checked that new and existing unit tests pass locally with
my changes

### Release Notes

- [x] Nodes (Validators and Full nodes): Add subscription to events
including filtering in gRPC

---------

Co-authored-by: Thoralf-M <46689931+Thoralf-M@users.noreply.github.com>
Co-authored-by: muXxer <git@muxxer.de>
Co-authored-by: Vlad Semenov <vlad.semenov@iota.org>

bingyanglin committed:
feat(gRPC): remain original checkpoint processing logic and enhance release_notes (#8371)

- Remain the execution order in `process_executed_checkpoint` as the
version before adding gRPC support
- Enhance `release_notes` rule to support `Internal gRPC API`
- Fix lint error due to rustc version update

No

- [x] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I have checked that new and existing unit tests pass locally with
my changes

Thoralf-M commented on crates/iota/src/iota_commands.rs:
```diff
@@ -27,6 +27,7 @@ use iota_genesis_builder::{SnapshotSource, SnapshotUrl};
 use iota_graphql_rpc::{
     config::ConnectionConfig, test_infra::cluster::start_graphql_server_with_fn_rpc,
 };
+use iota_grpc_api;
```
Useless import?

Thoralf-M replied:
Because there is another `Config` you use `iota_grpc_api::Config`so either remove this here or import the config with an alias

bingyanglin replied:
Thanks for pointing this out. Fixed at #8412

bingyanglin committed:
feat(gRPC): remove unused import (#8412)

# Description of change

Remove the unused import in `iota_commands.rs`

## Links to any relevant issues

No.

## How the change has been tested

- [x] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I have checked that new and existing unit tests pass locally with
my changes

bingyanglin commented:
Some changes after rebasing
- `cluster.wait_for_checkpoint(0, None).await;` is added to e2e testings, so we ensure the checkpoint executor is properly initialized and running before the gRPC streaming.
- We only need single `broadcast_checkpoint` for the latest `crates/iota-core/src/checkpoints/checkpoint_executor/mod.rs`, where we put `broadcast_checkpoint` after `bump_highest_executed_checkpoint()` in `async fn commit_checkpoint`.

semenov-vladyslav commented on crates/iota-grpc-api/tests/checkpoint_e2e.rs:
```diff
@@ -0,0 +1,193 @@
+
+    //  Ensures the checkpoint executor is properly initialized and running before
+    // the gRPC streaming
+    cluster.wait_for_checkpoint(0, None).await;
```
Doesn't look like a proper solution: client can cause server to panic. This should be fixed.

semenov-vladyslav commented on crates/iota-grpc-api/tests/checkpoint_e2e.rs:
```diff
@@ -0,0 +1,193 @@
+
+    //  Ensures the checkpoint executor is properly initialized and running before
+    // the gRPC streaming
+    cluster.wait_for_checkpoint(0, None).await;
```
Here too.

bingyanglin replied:
Thanks for pointing this out. As we discussed, fixed it in #8443.

semenov-vladyslav reviewed (changes requested):
Looks like the issue with "unfallible store methods".

bingyanglin committed:
fix: allow missing checkpoints in server init (#8443)

During server initialization it is expected that no checkpoints have
been executed yet. In the case we just return `None` to indicate service
is not ready rather than panicking.

None

- [x] Basic tests (linting, compilation, formatting, unit/integration
tests)
- [ ] Patch-specific tests (correctness, functionality coverage)
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I have checked that new and existing unit tests pass locally with
my changes

kodemartin reviewed (approved):
lgtm
